tinygrad/tensor.py

```diff
@@ -65,8 +65,9 @@ def _apply_winograd_matrix(mat, t:Tensor, dims:int) -> Tensor:
   ret = sum(prod(col[idx] for col, idx in zip(matcols, mat_is)) * t_[mat_is] for mat_is in itertools.product(range(len(mat[0])), repeat=dims))
   assert isinstance(ret, Tensor), "sum didn't return a Tensor"
   return ret
-
+@functools.lru_cache(maxsize=None)
```

Comment:
don't like these cache, probably clean up the underlying function to essential form, then think about caching

Reply:
The reason I added the caches here was because these functions are called over and over again with the same shapes.

for `test/external/external_test_speed_llama.py`

`_pad_left`

Without cache
```
Total time: 0.28224 s
File: /Users/rvd/src/rvd/tinygrad/tinygrad/tensor.py
Function: _pad_left at line 68

Line #      Hits         Time  Per Hit   % Time  Line Contents
==============================================================
    68                                           @profile
    69    103592     282240.0      2.7    100.0  def _pad_left(*shps:Tuple[sint, ...], v=1): return tuple((v,) * (max(len(i_) for i_ in shps) - len(i)) + i for i in shps)
```

With cache
```
Total time: 0.000599 s
File: /Users/rvd/src/rvd/tinygrad/tinygrad/tensor.py
Function: _pad_left at line 69

Line #      Hits         Time  Per Hit   % Time  Line Contents
==============================================================
    69                                           @functools.lru_cache(maxsize=None)
    70                                           @profile
    71       246        599.0      2.4    100.0  def _pad_left(*shps:Tuple[sint, ...], v=1): return tuple((v,) * (max(len(i_) for i_ in shps) - len(i)) + i for i in shps)
```


`_broadcast_shape`

Without cache:
```
Total time: 0.290469 s
File: /Users/rvd/src/rvd/tinygrad/tinygrad/tensor.py
Function: _broadcast_shape at line 70

Line #      Hits         Time  Per Hit   % Time  Line Contents
==============================================================
    70                                           @profile
    71     34302     290469.0      8.5    100.0  def _broadcast_shape(*shps:Tuple[sint, ...]): return tuple(0 if any(sh_ == 0 for sh_ in sh) else max(sh) for sh in zip(*_pad_left(*shps)))
```

With cache:
```
Total time: 0.000651 s
File: /Users/rvd/src/rvd/tinygrad/tinygrad/tensor.py
Function: _broadcast_shape at line 70

Line #      Hits         Time  Per Hit   % Time  Line Contents
==============================================================
    70                                           @functools.lru_cache(maxsize=None)
    71                                           @profile
    72        89        651.0      7.3    100.0  def _broadcast_shape(*shps:Tuple[sint, ...]): return tuple(0 if any(sh_ == 0 for sh_ in sh) else max(sh) for sh in zip(*_pad_left(*shps)))
```

Reply:
I think the reason why this was slow is because `max(len(i_) for i_ in shps)` recomputes for every `i`

caching without cleaning up the underlying code is just tech debt